### DISCUSSION TITLE
Upgrading Proxies from 3.0.0-beta to 4.0.1

### DISCUSSION THREAD

**Discussion by ccase47** [sebastienros/jint#1953] [Q&A] [answered]
Hi all,

I'm in the process of upgrading Jint from 3.0.0-beta-2033 to 4.0.1 and since ProxyInstance/JsProxy is now internal, I'm having some trouble trying to convert the logic we had at one point.

I want to access CLR objects as so -- `adoptablePets.dogs.Max`. I'm converting "adoptablePets" to my CLR object, and then trying to proxy the "Dogs" CLR object so the getter for it will find the dog named Max and return that.

In 3.0.0-beta we were able to do this:

```cs
private JsValue CreateDogsProxy(
	Engine engine,
	IConverter converter,
	IAdoptablePetCollection adoptablePets)
{
    var handler = new ObjectInstance(engine);
    
    JsValue GetterFunc(JsValue _, JsValue[] arguments)
    {
	    var name = (JsString)arguments[1];
	    var result = adoptablePets.Dogs.SingleOrDefault(v => v.Name == name;
	    if (result is null)
		    return JsValue.Undefined;
    
	    var dog = converter.Convert(engine, adoptablePets, result); // uses a converter to get the Dog object as JsValue
	    return dog;
    }
	    
    handler.Set("get", new ClrFunctionInstance(engine, "_get_", GetterFunc));
    
    return new ProxyInstance(engine, new ObjectInstance(engine), handler);
}
```
And then in JavaScript - `adoptablePets.dogs.Max` would return the correct Dog value.

Now since ProxyInstance/JsProxy is internal, I can't use this in the same way. I've tried creating my own proxy and also setting the "get" property for the object itself and nothing seems to hit the GetterFunc. What's the best way now to override the getter/setter for an object?

**Comment by lahma** [comment, marked as answer]
I've never thought Proxies being that convenient on interop side as there's already `ObjectWrapper`. Have you considered deriving from it and overriding relevant `Get` methods? Would probably be more performant too without all the abstractions.

**Reply by lahma** [reply]
Also with v4 a lot of things were made internal just to ensure easier refactoring and breaking things. They can always be opened again if/when needed.

**Reply by ccase47** [reply]
This did it for me! Thank you for this and for responding so quickly! I ended up creating a class that derives all methods from `ObjectWrapper`, and overrides the getter/setter with a handler's get and set properties if they exist to emulate a proxy.

**Reply by lahma** [reply]
Great to hear it worked! Customizing `ObjectWrapper` would be my general advice and skip the JS specific proxying as `ObjectWrapper` is the "more .NET native way of doing things".
